napari/layers/image/image.py

```diff
@@ -725,105 +728,76 @@ def _raw_to_displayed(self, raw):
-                / self.downsample_factors[self._thumbnail_level, not_disp]
-            )
-            downsampled_indices = np.round(
-                downsampled_indices.astype(float)
-            ).astype(int)
-            downsampled_indices = np.clip(
-                downsampled_indices,
-                0,
-                self.level_shapes[self._thumbnail_level, not_disp] - 1,
-            )
-            indices[not_disp] = downsampled_indices
+        # For the old experimental async code.
+        self._empty = False
 
-            thumbnail_source = self.data[self._thumbnail_level][tuple(indices)]
-        else:
-            self._transforms['tile2data'].scale = np.ones(self.ndim)
-            image_indices = self._slice_indices
-            image = self.data[image_indices]
-
-            # For single-scale we don't request a separate thumbnail_source
-            # from the ChunkLoader because in ImageSlice.chunk_loaded we
-            # call request.thumbnail_source() and it knows to just use the
-            # image itself is there is no explicit thumbnail_source.
-            thumbnail_source = None
-
-        # Load our images, might be sync or async.
-        data = self._SliceDataClass(
-            self, image_indices, image, thumbnail_source
+        # The new slicing code makes a request from the existing state and
+        # executes the request on the calling thread directly.
+        # For async slicing, the calling thread will not be the main thread.
+        request = self._make_slice_request_internal(self._slice_input)
+        response = request.execute()
+        self._set_slice_response(response, indices)
+
+    def _make_slice_request(self, dims: Dims) -> _ImageSliceRequest:
+        """Make an image slice request based on the given dims and this image."""
+        slice_input = self._make_slice_input(
+            dims.point, dims.ndisplay, dims.order
+        )
+        return self._make_slice_request_internal(slice_input)
+
+    def _make_slice_request_internal(
+        self, slice_input: _SliceInput
+    ) -> _ImageSliceRequest:
+        """Needed to support old-style sync slicing through set_view_slice."""
+        return _ImageSliceRequest(
+            dims=slice_input,
+            data=self.data,
+            # TODO: slice_indices should be lazily computed on the request itself,
+            # but this introduces some minor peformance issues right now related to
+            # evaluation of the data-to-world transform and its inverse.
+            slice_indices=self._slice_indices,
+            multiscale=self.multiscale,
+            corner_pixels=self.corner_pixels,
+            rgb=self.rgb,
+            data_level=self.data_level,
+            thumbnail_level=self._thumbnail_level,
+            level_shapes=self.level_shapes,
+            downsample_factors=self.downsample_factors,
+            lazy=True,
         )
-        self._load_slice(data)
+
+    def _set_slice_response(
```

Comment:
When I first skimmed through this implementation, I didn't realize what this function did because the name is a little confusing.

Reply:
Just to clarify the confusion, this made me think it was just a setter.

Reply:
In the future, it might just be a setter. Or it might go away entirely. But I appreciate that right now it does a lot more (i.e. there are a lot of side-effects). Would `_update_slice_response` be better? Any other ideas?

Reply:
`_update_slice_response` would be great!

Reply:
Done.